docs/design/variadics.md

```diff
@@ -0,0 +1,608 @@
+
+An expression of the form "`...,` _operand_" evaluates to a sequence of N
+values, where the k'th value is the value of _operand_ where `$I` is equal to
+k - 1.
+
+An expression of the form "`expand` _argument_" evaluates to "_argument_
+`[$I]`".
+
+An expression of the form "`each` _identifier_", where _identifier_ names a
+variadic binding, evaluates to the `$I`th value that it was bound to (indexed
+from zero).
+
+## Typechecking expressions and statements
+
+### Generalized tuple types
+
+The `...,` operator lets us form tuples out of sequences whose size is not known
+during typechecking. For example, in this code:
+
+```carbon
+fn F[each T:! type](x: (..., each i32), ..., each y: Optional(each T)) {
+  let z: auto = (..., expand x, 0 as f32, ..., each y);
+}
+```
+
+The type of `z` is a tuple whose elements are an indeterminate number of
+repetitions of `i32`, followed by `f32`, followed by a different indeterminate
+number of types that all have the form `Optional(T)` for some type `T`. We can't
+represent this as an explicit list of element types until those indeterminate
+numbers are known, so we need a more general representation for tuple types.
+
+In this model, a tuple type consists of a sequence of _pack components_, and a
+pack component consists of a type called the _representative_, and an arity,
+both of which may be symbolic. There is a special symbolic variable called the
+_pack index_, which can only be used in the representative of a pack component,
+and only as the subscript of an indexing expression on a tuple. The pack index
+is used solely as a placeholder, and never has a value. There are also symbolic
+variables representing the arity of every variadic binding. For purposes of
+illustration, the notation `<V, N>` represents a pack component with
+representative `V` and arity `N`, `$I` represents the pack index, and given a
+variadic binding `B`, `|B|` represents the arity of `B`, and `B/$I` represents
+the `$I`th value bound by `B`.
+
+So, continuing the earlier example, the type of `z` is represented symbolically
+as `(<i32, SizeOf(x)>, <f32, 1>, <Optional(T/$I), |T|>)`.
+
+A pack component is _variadic_ if its arity is unknown, and _singular_ if its
+arity is known to be 1 and its representative does not refer to the pack index.
+In contexts where all pack components are known to be singular, we will
+sometimes refer to them as "elements". Pack values are always assumed to be
+normalized, meaning that every component is either singular or variadic. This is
+always possible because if a component's arity is known to be some fixed value N
+other than 1, we can replace it with N singular components. The _shape_ of a
+tuple type is the sequence of arities of its components, so the shape of the
+type of `z` is `(SizeOf(x), 1, |T|)`.
+
+In order to index into a tuple, all of the components of its type must be
+singular, so that we can determine the type of the indexing expression.
```

Comment:
We could be a bit more permissive: in order to index into a tuple with index `I`, the shape of the tuple must begin with at least `I + 1` 1s, so that the `I`th element is known to be singular.